### PR TITLE
Add Track Changes feature

### PR DESCRIPTION
**Pull request by kluo**
* Creates "TrackChanges" django model, which ties a (submission_uuid,
  scorer_id) pair to a set of edits suggested by peers for a particular
  submission.
* Adds machinery for on-the-fly installation of the New York Times's ICE
  library for collaborative editing, allowing peer assessors to create
  marked-up versions of submissions, which can then be returned to the
  author as feedback.
* Adds machinery for displaying suggested edits to the creator of a
  submission with their scores and other feedback.
* Configurable within Studio on a per-Assessment basis by including a
  URL to the ICE library as part of the assessment definition.

### CODE REVIEW

File: openassessment/xblock/test/data/basic_scenario.xml [hunk 1]
```diff
@@ -88,6 +88,12 @@
     </rubric>
     <assessments>
         <assessment name="peer-assessment" must_grade="5" must_be_graded_by="3" />
+        <assessment
+            name="peer-assessment"
+            must_grade="5"
+            must_be_graded_by="3"
+            track_changes="http://example.org/track_changes.js"
```
Comment (stvstnfrd) [inline openassessment/xblock/test/data/basic_scenario.xml]: TODO: replace/update all instances like this.

File: openassessment/assessment/migrations/0004_add_trackchanges.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name='TrackChanges',
+            fields=[
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('owner_submission_uuid', models.UUIDField(db_index=True, default=uuid.uuid4)),
+                ('scorer_id', models.CharField(db_index=True, max_length=40)),
+                ('edited_content', models.TextField(blank=True)),
```
Comment (dcadams) [inline openassessment/assessment/migrations/0004_add_trackchanges.py]: I think you'll find this field is never used. It was the original storage place for edited content before adding capability of "parts". I kept it in as insurance to roll back should the parts change have caused chaos.

File: openassessment/xblock/openassessmentblock.py [hunk 3]
```diff
@@ -526,6 +532,17 @@ def _create_fragment(self, template, context_dict, initialize_js_func, additiona
...
             # minified additional_js should be already included in 'make javascript'
             fragment.add_javascript(load("static/js/openassessment-lms.min.js"))
+
+        ui_models = self._create_ui_models()
+        track_changes_fragments = [x['track_changes'] for x in ui_models if x.get('track_changes', None)]
+        if track_changes_fragments:
+            for tr_frag in track_changes_fragments:
+                fragment.add_javascript_url(tr_frag)  # TODO: move the URL to course advanced setting
+            if settings.DEBUG:
+                fragment.add_css_url(self.runtime.local_resource_url(self, "static/css/trackchanges.css"))
```
[the remark this answers is not in the document]
Comment (kluo) [reply openassessment/xblock/openassessmentblock.py]: I believe just following what upstream was doing: https://github.com/edx/edx-ora2/blame/master/openassessment/xblock/openassessmentblock.py#L511
Unbundled for local dev

**Comment by stvstnfrd** [comment]
Sorry @dcadams , I forgot about your work here too.
@kluo We should add David into the `Authored-by:` section!

**Comment by dcadams** [comment]
A couple of general points:
- You may want to consider, or think about, a site-wide and course-wide switch for this feature. May come up in the upstream review as would help usability.

- There's no JS tests. You may be lucky if they don't notice or it could go the other way.

File: openassessment/xblock/test/base.py [hunk 4]
```diff
@@ -332,6 +333,13 @@ def create_submission_and_assessments(
             # Create submissions and (optionally) assessments from other users
             must_be_graded_by = xblock.get_assessment_module('peer-assessment')['must_be_graded_by']
             scorer_subs = self.create_peer_submissions(student_item, peers, submission_text)
+
+            track_changes_edits = ''
+            if should_track_changes:
+                track_changes_edits = []
+                for sub_text in submission_text:
+                    track_changes_edits.append(sub_text + u'<span class="ins"> is wrong!</span>')
+
```
[the remark this answers is not in the document]
Comment (kluo) [reply openassessment/xblock/test/base.py]: This isn't quite the same though, this combines all the edits into one concatenated string, the code in question is produces a list with each edit as a separate element.

File: openassessment/xblock/static/js/src/studio/oa_edit_assessment.js [hunk 5]
```diff
@@ -120,6 +121,18 @@ OpenAssessment.EditPeerAssessmentView.prototype = {
         return this.mustBeGradedByField.get();
     },
 
+    trackChangesEnabled: function(isEnabled) {
+        var sel = $("#peer_assessment_track_changes", this.element);
```
[the remark this answers is not in the document]
Comment (kluo) [reply openassessment/xblock/static/js/src/studio/oa_edit_assessment.js]: Referring to the select element

**Comment by stvstnfrd** [comment]
Closing in favor of: https://github.com/edx/edx-ora2/pull/1258

**Pull request closed without merging**